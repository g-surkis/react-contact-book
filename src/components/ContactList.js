import React, { Component } from "react";
import PropTypes from "prop-types";
import { dropAction } from "../actions/actions";
import { connect } from "react-redux";
import ContactSticker from "./ContactSticker";

import PopUp from "./PopUp";

import { handleMoveCard } from "../services/contacts";

class ContactList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      edit: false,
      contacts: [],
      itemDragged: null
    };
    this.onEdit = this.onEdit.bind(this);
  }

  componentDidMount() {
    if (this.state.contacts.length === 0) {
      this.setState({ contacts: this.props.contacts });
    }
  }
  //for adding new contact just right after closed adding popup
  static getDerivedStateFromProps(props, state) {
    return {
      contacts: props.contacts
    };
  }
  onEdit() {
    this.setState({ edit: true });
  }
  itemDragged = el => {
    this.setState({
      itemDragged: el
    });
    return el;
  };
  moveCard = hoverIndex => {
    return handleMoveCard(
      this.state.contacts,
      this.state.itemDragged,
      hoverIndex,
      this.props.dropContact
    );
  };

  render() {
    const contactGrid = (
      <div>
        {this.state.contacts.map((element, i) => {
          return (
            <PopUp
              trigger={
                <div>
                  <ContactSticker
                    contact={element}
                    contactKey={i}
                    click={this.onEdit}
                    label={"edit"}
                    index={i}
                    moveCard={this.moveCard}
                    itemDragged={this.itemDragged}
                  />
                </div>
              }
              contact={element}
              contactKey={i}
              label={"Save Contact"}
              key={i}
            />
          );
        })}
      </div>
    );

    const addContactBtn = (
      <button className="animated_btns">Add Contact</button>
    );

    return (
      <div className="contacts_area">
        {contactGrid}

        <div className="btn_add_contact_area">
          <PopUp trigger={addContactBtn} label={"Add Contact"} />
        </div>
      </div>
    );
  }
}

function mapStateToProps(state) {
  return {
    contacts: state.contactsReducer
  };
}
function mapDispatchToProps(dispatch) {
  return {
    dropContact: obj => {
      dispatch(dropAction(obj));
    }
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ContactList);

ContactList.propTypes = {
  contacts: PropTypes.array,
  dropContacts: PropTypes.func
};
