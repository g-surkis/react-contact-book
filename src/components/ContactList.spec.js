import React from "react";
import { shallow, mount } from "enzyme";
import renderer from 'react-test-renderer';

import configureStore from 'redux-mock-store';

import ContactList from "./ContactList";

describe('>>>C O N T A C T  L I S T --- REACT-REDUX (Shallow + passing the {store} directly)',()=>{
    const initialState = { contactList: [{}] };
    const mockStore = configureStore();
    let store,container;

    beforeEach(()=>{
        store = mockStore(initialState);
        container = shallow(<ContactList store={store} /> );
    })

    it('+++ render the connected(SMART) component', () => {
       expect(container.length).toEqual(1);
    });

    it('+++ check Prop matches with initialState', () => {
       expect(container.prop('output')).toEqual(initialState.output);
    });


    describe('>>>C O N T A C T  L I S T --- Snapshot',()=>{
        it('+++capturing Snapshot of ContactList', () => {
            const renderedValue =  renderer.create(<ContactList output={10} store={store}/>).toJSON()
            expect(renderedValue).toMatchSnapshot();
        });
    });
})