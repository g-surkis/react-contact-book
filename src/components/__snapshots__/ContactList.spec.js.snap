// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>>C O N T A C T  L I S T --- REACT-REDUX (Shallow + passing the {store} directly) >>>C O N T A C T  L I S T --- Snapshot +++capturing Snapshot of ContactList 1`] = `
<div
  className="contacts_area"
>
  <div
    className="btn_add_contact_area"
  >
    <button
      className="animated_btns"
      onClick={[Function]}
    >
      Add Contact
    </button>
  </div>
</div>
`;
